### PR TITLE
feat(compiler-cli): Add an extended diagnostic for `ngSkipHydration`

### LINKED ISSUE #49501: Add extended template diagnostic to make sure `ngSkipHydration` is not a binding
### Which @angular/* package(s) are relevant/related to the feature request?

core

### Description

The `ngSkipHydration` is a special attribute that instructs Angular that hydration logic should be skipped for a given component, for example:
```
<cmp ngSkipHydration />
```
Angular expects this attribute to be known statically (i.e. it can not be a binding) and also it can not have values other than `"true"` or an empty value. This would be covered in documentation, but it'd be also great to add a template diagnostic to make sure that:
- a node doesn't have the `ngSkipHydration` in bindings
- the value of the `ngSkipHydration` is within the allowed range (`['true', '' /* empty string */]`)

The implementation would likely be very similar to [the `TextAttributeNotBindingSpec` check](https://github.com/angular/angular/blob/main/packages/compiler-cli/src/ngtsc/typecheck/extended/checks/text_attribute_not_binding/index.ts#L23).

Also related PR with a runtime check: https://github.com/angular/angular/pull/49500.

### PR DESCRIPTION
**Pull request by JeanMeche**
This diagnostic ensures that the special attribute `ngSkipHydration` is not a binding and has no other value than `"true"` or an empty value.

Fixes #49501

### CODE REVIEW

File: aio/content/extended-diagnostics/NG8108.md [hunk 1]
```diff
@@ -0,0 +1,29 @@
+@name ngSkipHydration biniding
+
+@description
+
+This diagnostic ensures that the special attribute `ngSkipHydration` is not a binding and has no 
+other value than `"true"` or an empty value. 
+
+
```
Comment (AndrewKushnir) [inline aio/content/extended-diagnostics/NG8108.md:8]:
```suggestion
The `ngSkipHydration` is a special attribute that indicates to Angular that a particular component should be
opted-out of hydration. This diagnostic ensures that this attribute `ngSkipHydration` is set statically and the 
value is either set to `"true"` or an empty value.

```

File: aio/content/extended-diagnostics/NG8108.md [hunk 2]
```diff
@@ -0,0 +1,29 @@
+@name ngSkipHydration biniding
```
Comment (AndrewKushnir) [inline aio/content/extended-diagnostics/NG8108.md:1]:
```suggestion
@name ngSkipHydration should be a static attribute
```

File: aio/content/extended-diagnostics/NG8108.md [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+@description
+
+This diagnostic ensures that the special attribute `ngSkipHydration` is not a binding and has no 
+other value than `"true"` or an empty value. 
+
+
+## What should I do instead?
+
+When using `ngSkipHydration`, ensure you do not use the Angular template binding syntax.
```
Comment (AndrewKushnir) [inline aio/content/extended-diagnostics/NG8108.md:11]:
```suggestion
When using the `ngSkipHydration`, ensure that it's set as a static attribute (i.e. you do not use the Angular template binding syntax).
```

File: aio/content/extended-diagnostics/NG8108.md [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+<pre>
+  <my-cmp ngSkipHydration></my-cmp>
+  <!-- or -->
+  <my-cmp ngSkipHydration="true"></my-cmp>
+</pre>
```
Comment (AndrewKushnir) [inline aio/content/extended-diagnostics/NG8108.md:17]: We can use the same approach as in other examples:
```suggestion
<code-example format="html" language="html">
&lt;my-cmp ngSkipHydration /&gt;
&lt;!-- or --&gt;
&lt;my-cmp ngSkipHydration="true" /&gt;
</code-example>
```

File: goldens/public-api/compiler-cli/extended_template_diagnostic_name.md [hunk 5]
```diff
@@ -17,6 +17,8 @@ export enum ExtendedTemplateDiagnosticName {
     // (undocumented)
     OPTIONAL_CHAIN_NOT_NULLABLE = "optionalChainNotNullable",
     // (undocumented)
+    SKIP_HYDRATION_BINDING = "skipHydrationBinding",
```
Comment (AndrewKushnir) [inline goldens/public-api/compiler-cli/extended_template_diagnostic_name.md:20]:
```suggestion
    SKIP_HYDRATION_NOT_STATIC = "skipHydrationNotStatic",
```

File: packages/compiler-cli/src/ngtsc/diagnostics/src/error_code.ts [hunk 6]
```diff
@@ -329,6 +329,21 @@ export enum ErrorCode {
    */
   OPTIONAL_CHAIN_NOT_NULLABLE = 8107,
 
+
+  /**
+   * `ngSkipHydration` should not be a binding .
```
Comment (AndrewKushnir) [inline packages/compiler-cli/src/ngtsc/diagnostics/src/error_code.ts:334]:
```suggestion
   * `ngSkipHydration` should not be a binding (it should be a static attribute).
```

File: packages/compiler-cli/src/ngtsc/typecheck/extended/checks/skip_hydration_binding/index.ts [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+class NgSkipHydrationSpec extends TemplateCheckWithVisitor<ErrorCode.SKIP_HYDRATION_BINDING> {
+  override code = ErrorCode.SKIP_HYDRATION_BINDING as const;
+
+  override visitNode(
+      ctx: TemplateContext<ErrorCode.SKIP_HYDRATION_BINDING>,
+      component: ts.ClassDeclaration,
+      node: TmplAstNode|AST,
+      ): NgTemplateDiagnostic<ErrorCode.SKIP_HYDRATION_BINDING>[] {
+    /** Binding should always error */
+    if ((node instanceof TmplAstBoundAttribute) && node.name === 'ngSkipHydration') {
```
Comment (AndrewKushnir) [inline packages/compiler-cli/src/ngtsc/typecheck/extended/checks/skip_hydration_binding/index.ts:29]: I think it'd be great to extract the name of the attribute into a separate const in this file. It'd help avoid typos.
```suggestion
    if ((node instanceof TmplAstBoundAttribute) && node.name === NG_SKIP_HYDRATION_ATTR_NAME) {
```

File: packages/compiler-cli/src/ngtsc/typecheck/extended/checks/skip_hydration_binding/index.ts [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+      const errorString = `ngSkipHydration should not be enclosed with square braces.`;
```
Comment (AndrewKushnir) [inline packages/compiler-cli/src/ngtsc/typecheck/extended/checks/skip_hydration_binding/index.ts:30]:
```suggestion
      const errorString = `ngSkipHydration should not be used as a binding.`;
```

**Review by AndrewKushnir — COMMENTED** [review]
@JeanMeche huge thanks for the help with this change! 👍 

The change looks great, I've just left a few minor comments.

File: packages/compiler-cli/src/ngtsc/typecheck/extended/checks/skip_hydration_not_static/index.ts [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+ */
+class NgSkipHydrationSpec extends TemplateCheckWithVisitor<ErrorCode.SKIP_HYDRATION_NOT_STATIC> {
+  override code = ErrorCode.SKIP_HYDRATION_NOT_STATIC as const;
+
+  override visitNode(
+      ctx: TemplateContext<ErrorCode.SKIP_HYDRATION_NOT_STATIC>, component: ts.ClassDeclaration,
+      node: TmplAstNode|AST): NgTemplateDiagnostic<ErrorCode.SKIP_HYDRATION_NOT_STATIC>[] {
+    /** Binding should always error */
+    if (node instanceof TmplAstBoundAttribute && node.name === NG_SKIP_HYDRATION_ATTR_NAME) {
+      const errorString = `ngSkipHydration should not be enclosed used as a binding.`;
```
Comment (AndrewKushnir) [inline packages/compiler-cli/src/ngtsc/typecheck/extended/checks/skip_hydration_not_static/index.ts:30]:
```suggestion
      const errorString = `ngSkipHydration should not be used as a binding.`;
```

**Review by AndrewKushnir — COMMENTED** [review]
@JeanMeche looks great, just a minor comment related to the error message text.

I also noticed that the commit message refers to the ApplicationRef (likely a commit message that was used in other PR). Could you please update the commit message?

Thank you.

**Review by AndrewKushnir — APPROVED** [review]
@JeanMeche looks great, thanks for the PR 👍

**Review by AndrewKushnir — APPROVED** [review]
Reviewed-for: public-api

**Comment by AndrewKushnir** [comment]
[Presubmit](http://test/OCL:518651004:BASE:518651027:1679515392767:782613b3).

**Review by jessicajaniuk — APPROVED** [review]
LGTM!

reviewed-for: public-api

**Comment by AndrewKushnir** [comment]
**Caretaker note**: presubmit is "green".

**Comment by atscott** [comment]
This PR was merged into the repository by commit 03d1d00ad9f88a2c449cceab64c1328787576162.

**Pull request closed without merging**